saltstack/salt
# Write lists of minions targeted by syndic masters to job cache

## Summary

When a master-of-masters publishes a command, and a syndic master re-publishes it to its subscribed minions, it fires an event back to the master containing the list of minions matched by the syndic master for that job.

This pull request scans these events for lists of minions, and writes them to the job cache.

Fixes #30528.
## How to test
1. Use [this documentation](https://help.github.com/articles/checking-out-pull-requests-locally/) to check out this pull request into a new branch, and install from git.
2. Recommended setup: 3 hosts (1 master-of-masters, 1 syndic master, 1 lower-level minion). Set `order_masters: True` on the master-of-masters and subscribe the lower-level minion to the syndic master, and the syndic master to the master-of-masters.
3. From the master-of-masters, run `salt -v '*' test.ping`
4. Take the jid obtained in step 3 and run `salt-run jobs.list_job <jid>`. The `Minions` list in the return data should contain the minion ID of the lower-level minion. Prior to this pull request, the lower-level minions would not be in this return data. After completing the rest of the test steps, installing Salt from the head of `2015.8` on the master-of-masters and repeating steps 3 and 4 should show no lower-level minions in the `jobs.list_job` return data.
5. Stop the `salt-minion` service on the lower-level minion, and then repeat step 3.
6. Take the jid obtained in step 5 and run `salt-run jobs.lookup_jid <jid> missing=True`. The lower-level minion should be included in the return data, and should show `Minion did not return`. Repeating the call to `jobs.lookup_jid` without the `missing=True` should exclude the lower-level minion from the return data.
## Review notes

Of particular focus in the review of this pull request should be the following:
- [ ] The `couchbase_return` returner. To maintain API compatibility I needed to move the writing of minion lists into a separate function, as I did for the `local_cache` returner, and I did not have an instance to test this.
- [ ] The file locking code added in salt/utils/files.py.
